### PR TITLE
[gpu] Replace volatile shared memory with shfl_sync in KNNSearch

### PR DESCRIPTION
**Pull request by haritha-j**
Fix for distance calculation issue caused by the use of volatile shared memory in GPU K nearest neighbour search.

### CODE REVIEW

**Comment by larshg** [comment]
These lines here:
https://github.com/PointCloudLibrary/pcl/pull/4306/files/49b80de604df8adac62cced748cbd0fe3fb2c188..b0ff2df0a61c5863c65a041466f2da26324f2224#diff-e1cd62e242c8446ecbfd699c09c5909fL181-L210
Can also be shared using the shfl instructions.

File: gpu/octree/src/cuda/knn_search.cu [hunk 1]
```diff
@@ -222,87 +222,50 @@ namespace pcl { namespace device { namespace knn_search
...
+					index = idx;
 				}
 			}
-			//parallel step
-			unsigned int lane = Warp::laneId();
 
-			float mind2 = dist2[tid];
+			//find minimum distance among warp threads
+            constexpr unsigned FULL_MASK = 0xFFFFFFFF;
+            static_assert(sizeof(KernelPolicy::WARP_SIZE) <= sizeof(unsigned int));
```
Comment (SergioRAgostinho) [inline gpu/octree/src/cuda/knn_search.cu:248]: this check is wrong. `WARP_SIZE` represents the number of bits in the mask. `sizeof` gives you the number of **bytes**. you also need an error message here explaining the user what is wrong if the static_assert is violated.

```suggestion
            static_assert(KernelPolicy::WARP_SIZE <= 8*sizeof(unsigned int));
```

File: gpu/octree/src/cuda/knn_search.cu [hunk 2]
```diff
@@ -222,87 +222,50 @@ namespace pcl { namespace device { namespace knn_search
...
 			{
-				float next = dist2[tid + 16];
-				if (mind2 > next) 
-				{ 
-					dist2[tid] = mind2 = next; 
-					index[tid] = index[tid + 16]; 
-				}                        
-			}
+				float next = __shfl_down_sync(FULL_MASK, dist2, bit_offset);
+				int next_index = __shfl_down_sync(FULL_MASK, index, bit_offset);
```
Comment (SergioRAgostinho) [inline gpu/octree/src/cuda/knn_search.cu:253]: const

File: gpu/octree/src/cuda/knn_search.cu [hunk 3]
```diff
@@ -222,87 +222,50 @@ namespace pcl { namespace device { namespace knn_search
...
-				float next = dist2[tid + 2];
-				if (mind2 > next) 
-				{ 
-					dist2[tid] = mind2 = next; 
-					index[tid] = index[tid + 2]; 
-				}                        
-			}
+			//retrieve index and dist2
+            index = __shfl_sync(FULL_MASK, index, 0);
+            dist2 = __shfl_sync(FULL_MASK, dist2, 0);
```
Comment (SergioRAgostinho) [inline gpu/octree/src/cuda/knn_search.cu:264]: no need for this temp.

File: gpu/octree/src/cuda/knn_search.cu [hunk 4]
```diff
@@ -222,87 +222,50 @@ namespace pcl { namespace device { namespace knn_search
...
+            dist = sqrt(dist2);
```
Comment (SergioRAgostinho) [inline gpu/octree/src/cuda/knn_search.cu:265]: const (in case it stops being an out parameter)

File: gpu/octree/src/cuda/knn_search.cu [hunk 5]
```diff
@@ -222,87 +222,50 @@ namespace pcl { namespace device { namespace knn_search
...
-				{ 
-					dist2[tid] = mind2 = next; 
-					index[tid] = index[tid + 1]; 
-				}                        
-			}        
-
-            dist = sqrt(dist2[tid - lane]);
-			return index[tid - lane];
-		}          
+			return index;
```
Comment (SergioRAgostinho) [inline gpu/octree/src/cuda/knn_search.cu:267]:
```suggestion
			return std::make_pair(index, dist);
```

**Comment by haritha-j** [comment]
I'm recieving the `pair` returned from `NearestWarpKernel` using auto here, which makes the next bit of code a bit less readable, but `std::tie` doesn't seem to work properly with cuda `__device__` functions, and `auto [offset, index]` is only supported from c++17 from what I read.

**Comment by kunaltyagi** [comment]
If the return type doesn't matter to the API/ABI (or explicitly needs to be given no API/ABI guarantee), you can use the following super-legal code
```cpp
auto myFunction(int a, double b) -> decltype(auto) {
  struct GuessTheReturnType {
    bool correct = false;
    float magic = 4.2e1;
  };
  return GuessTheReturnType{a, b};
}
```

**Comment by SergioRAgostinho** [comment]
@larshg please merge in case you're ok with the changes.

File: gpu/octree/src/cuda/knn_search.cu [hunk 6]
```diff
@@ -156,160 +157,107 @@ namespace pcl { namespace device { namespace knn_search
...
+
+            if (dist2 > d2) {
+              dist2 = d2;
+              index = idx;
+            }
+          }
+
+          // find minimum distance among warp threads
+          constexpr unsigned FULL_MASK = 0xFFFFFFFF;
+          static_assert(KernelPolicy::WARP_SIZE <= 8*sizeof(unsigned int));
```
Comment (larshg) [inline gpu/octree/src/cuda/knn_search.cu:235]: The static_assert without a string literal, requires c++17 in VS at least.
See: https://docs.microsoft.com/en-us/cpp/cpp/static-assert?view=vs-2019

**Comment by larshg** [comment]
https://github.com/PointCloudLibrary/pcl/blob/master/cmake/pcl_find_cuda.cmake

This one should probably also be revised soon. Not sure what cuda version is required for __shfl_sync instructions etc.

Right now it adds many cuda arch versions and with a new configure, it added 3.0 etch which failed to build.

File: gpu/octree/src/cuda/knn_search.cu (hunk 6 above)
Comment (SergioRAgostinho) [reply gpu/octree/src/cuda/knn_search.cu:235]: Hmm.. makes me wonder why didn't CI complain.

**Comment by SergioRAgostinho** [comment]
> https://github.com/PointCloudLibrary/pcl/blob/master/cmake/pcl_find_cuda.cmake
> 
> This one should probably also be revised soon. Not sure what cuda version is required for __shfl_sync instructions etc.
> 
> Right now it adds many cuda arch versions and with a new configure, it added 3.0 etch which failed to build.

Good catch. They were introduced in CUDA 9. CUDA C+14 support was also only introduced in CUDA 9. 

@PointCloudLibrary/maintainers any opposition to bump CUDA's minimum version to 9?

File: gpu/octree/src/cuda/knn_search.cu (hunk 6 above)
Comment (larshg) [reply gpu/octree/src/cuda/knn_search.cu:235]: We only have a linux GPU build :-)

**Comment by larshg** [comment]
Please go ahead and merge @SergioRAgostinho. I don't have those powers :smile:

**Pull request merged**